server/etcdserver/corrupt.go

```diff
@@ -258,57 +259,157 @@ func (cm *corruptionChecker) CompactHashCheck() {
+				zap.String("reason", reason))
+			continue
+		}
+
+		peersChecked++
+		if ids, ok := hashMap[peer.resp.Hash]; !ok {
+			hashMap[peer.resp.Hash] = []types.ID{peer.id}
+		} else {
+			ids = append(ids, peer.id)
+			hashMap[peer.resp.Hash] = ids
 		}
-		if len(peers) == peersChecked {
+	}
+
+	// All members have the same CompactRevision and Hash.
+	if len(hashMap) == 1 {
+		if peersChecked == len(peers) {
 			cm.lg.Info("successfully checked hash on whole cluster",
 				zap.Int("number-of-peers-checked", peersChecked),
-				zap.Int64("revision", hash.Revision),
+				zap.Int64("revision", leaderHash.Revision),
+				zap.Int64("compactRevision", leaderHash.CompactRevision),
 			)
 			cm.mux.Lock()
-			if hash.Revision > cm.latestRevisionChecked {
-				cm.latestRevisionChecked = hash.Revision
+			if leaderHash.Revision > cm.latestRevisionChecked {
+				cm.latestRevisionChecked = leaderHash.Revision
 			}
 			cm.mux.Unlock()
-			cm.lg.Info("finished compaction hash check", zap.Int("number-of-hashes-checked", i+1))
-			return
+			return true
 		}
 		cm.lg.Warn("skipped revision in compaction hash check; was not able to check all peers",
 			zap.Int("number-of-peers-checked", peersChecked),
 			zap.Int("number-of-peers", len(peers)),
-			zap.Int64("revision", hash.Revision),
+			zap.Int64("revision", leaderHash.Revision),
+			zap.Int64("compactRevision", leaderHash.CompactRevision),
 		)
+		// The only case which needs to check next hash
+		return false
 	}
-	cm.lg.Info("finished compaction hash check", zap.Int("number-of-hashes-checked", len(hashes)))
-	return
+
+	// Detected hashes mismatch
+	// The first step is to figure out the majority with the same hash.
+	memberCnt := len(peers) + 1
+	quorum := memberCnt/2 + 1
+	quorumExist := false
+	for k, v := range hashMap {
+		if len(v) >= quorum {
+			quorumExist = true
+			// remove the majority, and we might raise alarms for the left members.
+			delete(hashMap, k)
+			break
+		}
+	}
+
+	if !quorumExist {
+		// If quorumExist doesn't exist, then only raise alarm for the least minority
+		cm.lg.Error("Detected compaction hash mismatch but can't identify the corrupted members, so only raise alarm for the least minority",
+			zap.String("leader-id", leaderId.String()),
+			zap.Int64("leader-revision", leaderHash.Revision),
+			zap.Int64("leader-compact-revision", leaderHash.CompactRevision),
+			zap.Uint32("leader-hash", leaderHash.Hash),
+		)
+
+		// find out the members which are the least minority
+		var (
+			minCnt    int = math.MaxInt
+			hashVal   uint32
+			memberIDs types.IDSlice
+		)
+
+		for k, v := range hashMap {
+			if v.Len() < minCnt {
+				minCnt = v.Len()
+				hashVal = k
+				memberIDs = v
+			}
+		}
+
+		for _, pid := range memberIDs {
+			cm.hasher.TriggerCorruptAlarm(pid)
+		}
+		delete(hashMap, hashVal)
+
+		cm.lg.Error("Detected compaction hash mismatch but can't identify the corrupted members, so only raise alarm for the least minority",
```

Comment:
just wonder that we can merge this log with L356

```
cm.lg.Error("Detected compaction hash mismatch but can't identify the corrupted members, so only raise alarm for the least minority",
			zap.String("leader-id", leaderId.String()),
			zap.Int64("leader-revision", leaderHash.Revision),
			zap.Int64("leader-compact-revision", leaderHash.CompactRevision),
			zap.Uint32("leader-hash", leaderHash.Hash),
		)
```

It seems that L365 is duplicate.

Reply:
I forgot to remove the log at line 356. Removed.